### PR TITLE
Bonnie v2.1 VSAC Refactor UI Parameters

### PR DESCRIPTION
**Pull request by hossenlopp**
This changes the flow of VSAC parameters through measure loading so we have better control of what parameters go to VSAC. Some of the optional parameters were removed. VSAC username and password are no longer passed into measure loading code. Just a ticket granting ticket for VSAC.

Related bonnie PR: https://github.com/projecttacoma/bonnie/pull/974

Pull requests into Bonnie Bundler require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] This PR is into the correct branch.
- [x] JIRA ticket for this PR: https://jira.mitre.org/browse/BONNIE-1297 https://jira.mitre.org/browse/BONNIE-1350
- [x] JIRA ticket links to this PR
- [x] Code diff has been done and been reviewed (it **does not** contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass (remember to update Gemfile when applicable)
- [x] Code coverage has not gone down and all code touched or added is covered.

|             |  coverage                            |
| ------ | -------------------------- |
| before | 818 / 1287 LOC (63.56%) |
| after  | 807 / 1264 LOC (63.84%)    |

  All new or changed code is covered.

- [x] Automated regression test(s) pass N/A No change to calculation code.

If JIRA tests were used to supplement or replace automated tests: N/A
- [x] JIRA test links: N/A
- [x] Justification for using JIRA tests: N/A
- [x] JIRA tests have been added to sprint N/A

**Reviewer 1:** @mayerm94

Name:
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests: N/A
- [x] JIRA tests have been run and pass N/A
- [x] You agree with the justification for use of JIRA tests or have provided input on why you disagree N/A

**Reviewer 2:**

Name:
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests: N/A
- [x] JIRA tests have been run and pass N/A
- [x] You agree with the justification for use of JIRA tests or have provided input on why you disagree N/A

### CODE REVIEW

File: lib/measures/loading/value_set_loader.rb [hunk 1]
```diff
@@ -19,125 +19,95 @@ def self.get_value_set_models(value_set_oids, user=nil)
...
+          if vs_element && vs_element['ID'] == value_set[:oid]
+            vs_element['id'] = value_set[:oid]
+            set = HealthDataStandards::SVS::ValueSet.load_from_xml(doc)
+            # make sure this value set has concepts, delete it and raise error if it is empty
+            if set.concepts.empty?
+              set.delete
+              raise Util::VSAC::VSEmptyError.new(value_set[:oid])
+            end
+            set.user = user
+            #bundle id for user should always be the same 1 user to 1 bundle
```
Comment (mayerm94) [inline lib/measures/loading/value_set_loader.rb]: Space after start of comment

File: lib/measures/loading/value_set_loader.rb [hunk 2]
```diff
@@ -19,125 +19,95 @@ def self.get_value_set_models(value_set_oids, user=nil)
...
-          backup_vs = get_existing_vs(user, value_set_oids).to_a
-          delete_existing_vs(user, value_set_oids)
+
+      errors = {}
+      api = Util::VSAC::VSACAPI.new(config: APP_CONFIG['vsac'], ticket_granting_ticket: vsac_ticket_granting_ticket)
+
+      RestClient.proxy = ENV["http_proxy"]
+      value_sets.each do |value_set|
+        # The vsac_options that will be used for this specific value set
+        vs_vsac_options = nil
```
Comment (mayerm94) [inline lib/measures/loading/value_set_loader.rb]: The following logic could be simplified by setting vs_vsac_options to vsac_options by default here, so you don't need to set it in both of the else clauses as a fall-back

File: lib/measures/loading/value_set_loader.rb [hunk 3]
```diff
@@ -19,125 +19,95 @@ def self.get_value_set_models(value_set_oids, user=nil)
...
         end
-        vsac_options = { config: APP_CONFIG['vsac'] }
-        if ticket_granting_ticket != nil
-          # we have to assume expiration time for now
-          vsac_options[:ticket_granting_ticket] = ticket_granting_ticket
-        elsif username != nil
-          vsac_options[:username] = username
-          vsac_options[:password] = password
+
+        # Determine version to store value sets as after parsing. left nil if this is not needed
```
Comment (mayerm94) [inline lib/measures/loading/value_set_loader.rb]: Left empty, not nil

File: lib/measures/loading/value_set_loader.rb [hunk 4]
```diff
@@ -19,125 +19,95 @@ def self.get_value_set_models(value_set_oids, user=nil)
...
+            # make sure this value set has concepts, delete it and raise error if it is empty
+            if set.concepts.empty?
+              set.delete
+              raise Util::VSAC::VSEmptyError.new(value_set[:oid])
+            end
+            set.user = user
+            #bundle id for user should always be the same 1 user to 1 bundle
+            #using this to allow cat I generation without extensive modification to HDS
+            set.bundle = user.bundle if (user && user.respond_to?(:bundle))
+            # As of t9/7/2017, when valuesets are retrieved from VSAC via profile, their version defaults to N/A
```
Comment (mayerm94) [inline lib/measures/loading/value_set_loader.rb]: Remove extra 't' before date

File: lib/measures/loading/value_set_loader.rb [hunk 5]
```diff
@@ -19,125 +19,95 @@ def self.get_value_set_models(value_set_oids, user=nil)
...
-        if use_cache
-          codeset_base_dir = Measures::Loader::VALUE_SET_PATH
-          FileUtils.mkdir_p(codeset_base_dir)
+        # only access the database if we don't intend on using cached values
+        set = HealthDataStandards::SVS::ValueSet.where({user_id: user.id, oid: value_set[:oid], version: query_version}).first()
+        if (vs_vsac_options[:include_draft] && set)
+          set.delete
+          set = nil
         end
+        # TODO: figure out if this is still a good idea
```
Comment (mayerm94) [inline lib/measures/loading/value_set_loader.rb]: Is this TODO still relevant? It makes sense to me to me to add the set to the existing_value_set_map here.

File: lib/measures/loading/value_set_loader.rb [hunk 6]
```diff
@@ -19,125 +19,95 @@ def self.get_value_set_models(value_set_oids, user=nil)
...
-          # only access the database if we don't intend on using cached values
-          set = HealthDataStandards::SVS::ValueSet.where({user_id: user.id, oid: value_set[:oid], version: query_version}).first() unless use_cache
-          if (includeDraft && set)
-            set.delete
-            set = nil
-          end
-          if (set)
+          vs_data = api.get_valueset(value_set[:oid], vs_vsac_options)
+
+          vs_data.force_encoding("utf-8") # there are some funky unicodes coming out of the vs response that are not in ASCII as the string reports to be
```
Comment (mayerm94) [inline lib/measures/loading/value_set_loader.rb]: Move comment above line, this line is long

File: lib/measures/loading/value_set_loader.rb [hunk 7]
```diff
@@ -19,125 +19,95 @@ def self.get_value_set_models(value_set_oids, user=nil)
...
+        set = HealthDataStandards::SVS::ValueSet.where({user_id: user.id, oid: value_set[:oid], version: query_version}).first()
+        if (vs_vsac_options[:include_draft] && set)
+          set.delete
+          set = nil
         end
+        # TODO: figure out if this is still a good idea
+        if (set)
+          existing_value_set_map[set.oid] = set
+        else
+          vs_data = nil
```
Comment (mayerm94) [inline lib/measures/loading/value_set_loader.rb]: I think this line can be deleted

**Review by mayerm94 — DISMISSED** [review]

File: lib/measures/loading/value_set_loader.rb [hunk 8]
```diff
@@ -19,125 +19,95 @@ def self.get_value_set_models(value_set_oids, user=nil)
...
+        elsif vs_vsac_options[:release]
+          query_version = vs_vsac_options[:release]
         end
-        errors = {}
-        api = Util::VSAC::VSACAPI.new(vsac_options)
 
-        if use_cache
-          codeset_base_dir = Measures::Loader::VALUE_SET_PATH
-          FileUtils.mkdir_p(codeset_base_dir)
+        # only access the database if we don't intend on using cached values
```
Comment (jbradl11) [inline lib/measures/loading/value_set_loader.rb]: Can this comment be removed now that we are no longer checking for `use_cache`?

File: lib/measures/loading/value_set_loader.rb (hunk 1 above)
Comment (hossenlopp) [reply lib/measures/loading/value_set_loader.rb]: Existing code issue. I'll fix it though.

File: lib/measures/loading/value_set_loader.rb (hunk 2 above)
Comment (hossenlopp) [reply lib/measures/loading/value_set_loader.rb]: Good call. I'll adjust this logic and comments.

File: lib/measures/loading/value_set_loader.rb (hunk 4 above)
Comment (hossenlopp) [reply lib/measures/loading/value_set_loader.rb]: Existing code issue. I'll fix it though.

File: lib/measures/loading/value_set_loader.rb (hunk 7 above)
Comment (hossenlopp) [reply lib/measures/loading/value_set_loader.rb]: Haha yep!

File: lib/measures/loading/value_set_loader.rb (hunk 3 above)
Comment (hossenlopp) [reply lib/measures/loading/value_set_loader.rb]: Will fix comment.

File: lib/measures/loading/value_set_loader.rb (hunk 6 above)
Comment (hossenlopp) [reply lib/measures/loading/value_set_loader.rb]: Existing code issue. I'll fix it though.

File: lib/measures/loading/value_set_loader.rb (hunk 4 above)
Comment (hossenlopp) [reply lib/measures/loading/value_set_loader.rb]: Existing code issue. I'll fix it though.

File: lib/measures/loading/value_set_loader.rb (hunk 6 above)
Comment (hossenlopp) [reply lib/measures/loading/value_set_loader.rb]: Existing code issue. I'll fix it though.

File: lib/measures/loading/value_set_loader.rb (hunk 7 above)
Comment (hossenlopp) [reply lib/measures/loading/value_set_loader.rb]: Yep! Will remove it.

File: lib/measures/loading/value_set_loader.rb (hunk 8 above)
Comment (hossenlopp) [reply lib/measures/loading/value_set_loader.rb]: I'll update it to be more accurate to what's actually going on.

**Review by jbradl11 — APPROVED** [review]
Code and tests look good. Tests pass.

**Review by mayerm94 — APPROVED** [review]
Newest changes look good.
